File: src/vllm-sr/cli/evaluation/campaign_protocol.py
"""Versioned catalog descriptor for Campaign source cohorts."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Literal

from pydantic import Field

from cli.evaluation.contract_primitives import StrictModel
from cli.evaluation.execution_contract import (
    LIVE_RUNTIME_EXECUTOR_ID,
    MOM_REPLAY_EXECUTOR_ID,
)
from cli.evaluation.reporting import EvidenceLevel, TrackID

CAMPAIGN_COHORT_SCHEMA_VERSION = "evaluation-campaign-cohort.v1"


class CampaignProtocol(StrictModel):
    """Declares that a suite can supply one Campaign comparison cohort."""

    schema_version: Literal[CAMPAIGN_COHORT_SCHEMA_VERSION] = (
        CAMPAIGN_COHORT_SCHEMA_VERSION
    )
    minimum_cases: int = Field(gt=0, strict=True)


def validate_campaign_protocol(
    protocol: CampaignProtocol | None,
    *,
    modes: tuple[Literal["replay", "live"], ...],
    executors: Mapping[Literal["replay", "live"], str],
    track_ids: tuple[TrackID, ...],
    evidence_level: EvidenceLevel,
    case_count: int | None,
) -> None:
    """Validate a present descriptor against only its containing suite."""

    if protocol is None:
        return
    if (
        modes != ("replay", "live")
        or executors
        != {
            "replay": MOM_REPLAY_EXECUTOR_ID,
            "live": LIVE_RUNTIME_EXECUTOR_ID,
        }
        or track_ids != ("routing", "model_pool", "joint")
        or evidence_level != "E0"
        or case_count is None
        or case_count <= 0
        or protocol.minimum_cases > case_count
    ):
        raise ValueError(
            "campaign protocol requires a valid E0 replay/live MoM cohort contract"
        )

File: dashboard/backend/evaluationplane/campaign_protocol.go
package evaluationplane

import (
	"errors"
	"reflect"
)

const campaignCohortSchemaVersion = "evaluation-campaign-cohort.v1"

// CampaignProtocol declares that a catalog suite can supply one Campaign
// comparison cohort. Its presence is the eligibility signal.
type CampaignProtocol struct {
	SchemaVersion string `json:"schema_version"`
	MinimumCases  int    `json:"minimum_cases"`
}

func validateCampaignProtocol(suite CatalogSuite) error {
	protocol := suite.CampaignProtocol
	if protocol == nil {
		return nil
	}
	requiredExecutors := map[Mode]string{
		ModeReplay: momReplayExecutorID,
		ModeLive:   liveRuntimeExecutorID,
	}
	if protocol.SchemaVersion != campaignCohortSchemaVersion || protocol.MinimumCases <= 0 ||
		!reflect.DeepEqual(suite.Modes, []Mode{ModeReplay, ModeLive}) ||
		!reflect.DeepEqual(suite.Executors, requiredExecutors) ||
		!reflect.DeepEqual(suite.TrackIDs, []TrackID{"routing", "model_pool", "joint"}) ||
		suite.EvidenceLevel != "E0" || suite.CaseCount <= 0 || protocol.MinimumCases > suite.CaseCount {
		return errors.New("campaign protocol is inconsistent with its evaluation suite")
	}
	return nil
}
